cloud/scope/powervs_machine_test.go

```diff
@@ -122,74 +133,1188 @@ func newDHCPServer(serverID, networkID string) models.DHCPServers {
+			}
+
+			mockpowervs.EXPECT().GetAllNetwork().Return(networks, nil)
+			scope := PowerVSMachineScope{
+				IBMPowerVSClient: mockpowervs,
+			}
+			networkID, err := getNetworkID(networkResource, &scope)
+			g.Expect(*networkID).To(Equal(expectedNetworkID))
+			g.Expect(err).To(BeNil())
+		})
+
+		t.Run("Failed to fetch network ID with matching regex", func(t *testing.T) {
+			g := NewWithT(t)
+			setup(t)
+			t.Cleanup(teardown)
+			expectedNetworkID := "foo-netID"
+			regex := "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
+			networks := &models.Networks{
+				Networks: []*models.NetworkReference{
+					{
+						Name: core.StringPtr(expectedNetworkID),
+					},
+				},
+			}
+			networkResource := infrav1beta2.IBMPowerVSResourceReference{
+				RegEx: core.StringPtr(regex),
+			}
+
+			mockpowervs.EXPECT().GetAllNetwork().Return(networks, nil)
+			scope := PowerVSMachineScope{
+				IBMPowerVSClient: mockpowervs,
+			}
+			networkID, err := getNetworkID(networkResource, &scope)
+			g.Expect(networkID).To(BeNil())
+			g.Expect(err.Error()).To(Equal(fmt.Sprintf("failed to find a network ID with RegEx %s", regex)))
+		})
+
+		t.Run("When ID, name and regex are all nil", func(t *testing.T) {
+			g := NewWithT(t)
+			networkResource := infrav1beta2.IBMPowerVSResourceReference{}
+			scope := PowerVSMachineScope{}
+			networkID, err := getNetworkID(networkResource, &scope)
+			g.Expect(networkID).To(BeNil())
+			g.Expect(err.Error()).To(Equal("ID, Name and RegEx can't be nil"))
+		})
+	})
+}
+
+func TestGetMachineInternalIP(t *testing.T) {
+	t.Run("Get Machine Internal IP", func(t *testing.T) {
+		t.Run("Returns machine IP for address type - Node Internal IP", func(t *testing.T) {
+			g := NewWithT(t)
+			expectedAddress := "10.0.0.1"
+			scope := PowerVSMachineScope{
+				IBMPowerVSMachine: &infrav1beta2.IBMPowerVSMachine{
+					Status: infrav1beta2.IBMPowerVSMachineStatus{
+						Addresses: []corev1.NodeAddress{
+							{
+								Type:    corev1.NodeInternalIP,
+								Address: expectedAddress,
+							},
+						},
+					},
+				},
+			}
+			g.Expect(expectedAddress).To(Equal(scope.GetMachineInternalIP()))
+		})
+
+		t.Run("Returns empty IP for address type - node external IP", func(t *testing.T) {
+			g := NewWithT(t)
+			scope := PowerVSMachineScope{
+				IBMPowerVSMachine: &infrav1beta2.IBMPowerVSMachine{
+					Status: infrav1beta2.IBMPowerVSMachineStatus{
+						Addresses: []corev1.NodeAddress{
+							{
+								Type:    corev1.NodeExternalIP,
+								Address: "198.0.0.1",
+							},
+						},
+					},
+				},
+			}
+			g.Expect("").To(Equal(scope.GetMachineInternalIP()))
+		})
+
+		t.Run("Returns empty IP if powervsmachineStatus in nil", func(t *testing.T) {
+			g := NewWithT(t)
+			scope := PowerVSMachineScope{
+				IBMPowerVSMachine: &infrav1beta2.IBMPowerVSMachine{},
+			}
+			g.Expect("").To(Equal(scope.GetMachineInternalIP()))
+		})
+	})
+}
+
+func TestSetProviderID(t *testing.T) {
```

Comment:
Recently added other case where name is set we fetch from cloud https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/2039 , Can we cover that case as well?